gate-core/src/main/groovy/com/netflix/spinnaker/gate/security/RequestContext.java

```diff
@@ -16,61 +16,71 @@
 
 package com.netflix.spinnaker.gate.security;
 
+import com.netflix.spinnaker.security.AuthenticatedRequest;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * No longer backed by a thread local, this is now a readonly proxy to {@link AuthenticatedRequest},
+ * which is the only source of truth for context values. This is to avoid the situation where
+ * RequestContext and {@link AuthenticatedRequest} provide an inconsistent view of the context or
+ * lose it across thread boundaries.
+ */
+@Slf4j
 public class RequestContext {
-  private static final ThreadLocal<RequestContext> threadLocal = new ThreadLocal<>();
+  private static final RequestContext placeholder = new RequestContext();
 
-  private String application;
-  private String authenticatedUser;
-  private String executionType;
-  private String executionId;
-  private String origin;
+  private RequestContext() {}
 
   public RequestContext(String origin, String authenticatedUser) {
-    this.origin = origin;
-    this.authenticatedUser = authenticatedUser != null ? authenticatedUser : "anonymous";
+    log.warn("call to deprecated RequestContext's constructor");
   }
 
+  @Deprecated
   public static void set(RequestContext requestContext) {
-    threadLocal.set(requestContext);
+    log.warn("call to deprecated method RequestContext::set");
```

Comment:
this trick proved useful to hunt down callers (https://github.com/spinnaker/kork/blob/master/kork-web/src/main/java/com/netflix/spinnaker/okhttp/MetricsInterceptor.java#L93-L97)